chisel/src/config.rs

```diff
@@ -0,0 +1,570 @@
+            modules: module_configs,
+        })
+    }
+}
+
+impl FromYaml for ChiselConfig {
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        // Ensure that the first value in the config is a mapping representing rulesets.
+        if !yaml.is_mapping() {
+            return Err("Top-level value is not a mapping".to_string());
+        }
+
+        let rulesets = yaml.as_mapping().expect("Already validated");
+
+        // Ensure that all rulesets in the top-level YAML map are string-map pairs.
+        // Valid example:
+        // ruleset:
+        //   option: value
+        //
+        // Invalid example:
+        // ruleset: "something"
+        if rulesets
+            .iter()
+            .find(|(key, val)| !key.is_string() || !val.is_mapping())
+            .is_some()
+        {
+            return Err(
+                "Malformed ruleset; expected all rulesets to be string-map pairs".to_string(),
+            );
+        }
+
+        let rulesets: VecDeque<(String, Ruleset)> = rulesets
+            .iter()
+            .map(|(key, val)| {
+                (
+                    key.as_str().expect("Already validated").to_string(),
+                    Ruleset::from_yaml(val).expect("No failure cases yet"),
+                )
+            })
+            .collect();
+
+        Ok(ChiselConfig(rulesets))
+    }
+}
+
+impl FromYaml for Ruleset {
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let mut opts: HashMap<String, String> = HashMap::new();
+        let mut modules: VecDeque<(String, ModuleConfig)> = VecDeque::new();
+
+        let ruleset = yaml.as_mapping().expect("Already validated");
+        for (key, val) in ruleset.iter() {
+            match (key, val) {
+                // If string-string pair, treat it as an option.
+                (Value::String(k), Value::String(v)) => {
+                    opts.insert(k.to_string(), v.to_string());
+                }
+                // If string-map pair or a string-null pair (no special options), treat it as a module configuration.
+                (Value::String(k), v) => {
+                    if v.is_mapping() || v.is_null() {
+                        modules.push_back((
+                            k.to_string(),
+                            ModuleConfig::from_yaml(v).expect("No failure cases yet"),
+                        ));
+                    }
+                }
+                // ignore others for now
+                (_, _) => (),
+            }
+        }
+        Ok(Ruleset { opts, modules })
+    }
+}
+
+impl FromYaml for ModuleConfig {
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        if yaml == &Value::Null {
+            Ok(ModuleConfig(HashMap::new()))
+        } else {
+            let opts = yaml.as_mapping().expect("Already validated");
+            let opts: HashMap<String, String> = opts
```

Comment:
If these two lines are merged is rustfmt breaking it up weirdly?

Reply:
No, I just thought it would be more clear to assign `opts` in two expressions.